order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java

```diff
@@ -1,58 +1,179 @@
+        } catch (InvalidProtocolBufferException e) {
+            throw new IncorrectResponseException("Response is not correct!");
+        }
+        Set<Item> itemSet = this.allItems.get(tickResponse.getGoodName());
+        if (itemSet == null) {
+            LOGGER.error("Item: {} is not being tracked and cannot be added to itemMarket!",
+                    tickResponse.getGoodName());
+            return;
+        }
+        Item item = populateItem(tickResponse);
+
+        this.productsQuantity.merge(tickResponse.getGoodName(), tickResponse.getQuantity(), Long::sum);
+
+        LOGGER.info("Products data updated!");
+
+        if (itemSet.contains(item)) {
+            Item presentItem = itemSet.stream().filter(currentItem -> Double.compare(currentItem.getPrice(), item.getPrice()) == 0
+                    && currentItem.getOrigin().equals(item.getOrigin())).findFirst().get();
+            presentItem.setQuantity(presentItem.getQuantity() + item.getQuantity());
+            return;
+        }
+
+        itemSet.add(item);
+
+    }
+
+    public List<OrderBookLayer> getRequestedItem(String itemName, long quantity) {
+        LOGGER.info("Getting requested item: {} with quantity: {}", itemName, quantity);
+        TreeSet<Item> items = this.allItems.get(itemName);
+
+        if (items == null || this.productsQuantity.get(itemName) < quantity) {
+            return Collections.emptyList();
+        }
+
+        List<OrderBookLayer> layers;
+        try {
+            lock.writeLock().lock();
+
+            layers = new ArrayList<>();
+
+            Iterator<Item> iterator = items.iterator();
+            long itemLeft = quantity;
+
+            while (iterator.hasNext() && itemLeft > 0) {
+                Item currentItem = iterator.next();
+
+                OrderBookLayer.Builder currentLayer = populateItemLayer(iterator, itemLeft, currentItem);
+
+                this.productsQuantity.put(itemName, this.productsQuantity.get(itemName) - currentLayer.getQuantity());
+                itemLeft -= currentLayer.getQuantity();
+
+                OrderBookLayer orderBookLayer = currentLayer
+                        .setOrigin(currentItem.getOrigin())
+                        .build();
+                layers.add(orderBookLayer);
+            }
+        } finally {
+            lock.writeLock().unlock();
+        }
+        return layers;
+    }
+
+    private OrderBookLayer.Builder populateItemLayer(Iterator<Item> iterator, long itemLeft, Item currentItem) {
+        OrderBookLayer.Builder currentLayer = OrderBookLayer.newBuilder()
+                .setPrice(currentItem.getPrice());
+
+        if (currentItem.getQuantity() >= itemLeft) {
+            currentLayer.setQuantity(itemLeft);
+
+            if (currentItem.getQuantity() == itemLeft) {
+                iterator.remove();
+            }
+
+            currentItem.setQuantity(currentItem.getQuantity() - itemLeft);
+        } else if (currentItem.getQuantity() < itemLeft) {
```

Comment:
the if is not necessary here, the first if is checking if it is >= then the else will be executed for <

Reply:
Fixed